csharp/ql/src/experimental/CWE-918/RequestForgery.qll

```diff
@@ -0,0 +1,128 @@
+
+    override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+      exists(Constructor c | c.getDeclaringType().hasQualifiedName("System.Uri") |
+        c.getParameter(0) = node1.asParameter() and c.getACall() = node2.asExpr()
+      )
+    }
+
+    override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
+      guard instanceof SanitizerGuard
+    }
+  }
+
+  /**
+   * A remote data flow source taken as a source
+   * for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class RemoteFlowSourceAsSource extends Source {
+    RemoteFlowSourceAsSource() { this instanceof RemoteFlowSource }
+  }
+
+  /**
+   * An url argument to a `HttpRequestMessage` constructor call
+   * taken as a sink for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class SystemWebHttpRequestMessageSink extends Sink {
+    SystemWebHttpRequestMessageSink() {
+      exists(Class c | c.hasQualifiedName("System.Net.Http.HttpRequestMessage") |
+        c.getAConstructor().getACall().getArgument(1) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An argument to a `WebRequest.Create` call taken as a
+   * sink for Server Side Request Forgery(SSRF) Vulnerabilities. *
+   */
+  private class SystemNetWebRequestCreateSink extends Sink {
+    SystemNetWebRequestCreateSink() {
+      exists(Method m |
+        m.getDeclaringType().hasQualifiedName("System.Net.WebRequest") and m.hasName("Create")
+      |
+        m.getACall().getArgument(0) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An argument to a new HTTP Request call of a `System.Net.Http.HttpClient` object
+   * taken as a sink for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class SystemNetHttpClientSink extends Sink {
+    SystemNetHttpClientSink() {
+      exists(Method m |
+        m.getDeclaringType().hasQualifiedName("System.Net.Http.HttpClient") and
+        m.hasName([
+            "DeleteAsync", "GetAsync", "GetByteArrayAsync", "GetStreamAsync", "GetStringAsync",
+            "PatchAsync", "PostAsync", "PutAsync"
+          ])
+      |
+        m.getACall().getArgument(0) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An url argument to a method call of a `System.Net.WebClient` object
+   * taken as a sink for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class SystemNetClientBaseAddressSink extends Sink {
+    SystemNetClientBaseAddressSink() {
+      exists(Property p |
+        p.hasName("BaseAddress") and
+        p.getDeclaringType()
+            .hasQualifiedName(["System.Net.WebClient", "System.Net.Http.HttpClient"])
+      |
+        p.getAnAssignedValue() = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * A method call which checks the base of the tainted uri is assumed
+   * to be a guard for Server Side Request Forgery(SSRF) Vulnerabilities.
+   * This guard considers all checks as valid.
+   */
+  private class BaseUriGuard extends SanitizerGuard, MethodCall {
```

Comment:
I think there could/should be other sanitizers too. The base URL could be checked simply through a string operation.